File: apps/android/src/main/kotlin/com/sierraespada/wakeywakey/billing/EntitlementManager.kt
package com.sierraespada.wakeywakey.billing

import android.content.Context
import com.revenuecat.purchases.kmp.Purchases
import com.revenuecat.purchases.kmp.models.CustomerInfo
import com.revenuecat.purchases.kmp.models.Offerings
import com.revenuecat.purchases.kmp.models.Package
import com.revenuecat.purchases.kmp.models.PurchasesError
import com.revenuecat.purchases.kmp.models.StoreTransaction
import com.sierraespada.wakeywakey.analytics.AnalyticsProvider
import com.sierraespada.wakeywakey.analytics.Event
import com.sierraespada.wakeywakey.settings.SettingsRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.concurrent.TimeUnit

/**
 * Gestiona el acceso Pro del usuario combinando dos fuentes:
 *
 *  1. Trial de app (14 días desde primera instalación, sin pedir nada)
 *  2. Suscripción o compra única via RevenueCat / Play Store
 *
 * [isPro] = trial activo OR suscripción activa
 */
object EntitlementManager {

    private const val ENTITLEMENT_ID = "WakeyWakey Pro"
    const val TRIAL_DAYS = 14L

    private val scope = CoroutineScope(Dispatchers.IO)

    // ── Estado ────────────────────────────────────────────────────────────────

    /** Días restantes del trial de app (0 = expirado o ya es suscriptor) */
    private val _trialDaysLeft = MutableStateFlow(TRIAL_DAYS.toInt())
    val trialDaysLeft: StateFlow<Int> = _trialDaysLeft.asStateFlow()

    /** true si tiene suscripción/compra activa en RevenueCat */
    private val _isRevenueCatPro = MutableStateFlow(false)

    /** true si trial activo O suscripción activa */
    private val _isPro = MutableStateFlow(true) // true por defecto hasta que init() compruebe
    val isPro: StateFlow<Boolean> = _isPro.asStateFlow()

    private val _offerings = MutableStateFlow<Offerings?>(null)
    val offerings: StateFlow<Offerings?> = _offerings.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    // ── Init ──────────────────────────────────────────────────────────────────

    /**
     * Llamar en Application.onCreate() con el contexto.
     * Calcula los días de trial restantes y carga el estado de RevenueCat.
     */
    fun init(context: Context) {
        scope.launch {
            val repo        = SettingsRepository.getInstance(context)
            val isFirstRun  = repo.getInstallDate() == null
            val installDate = repo.getOrSetInstallDate()
            val daysSince   = TimeUnit.MILLISECONDS.toDays(System.currentTimeMillis() - installDate)
            val daysLeft    = (TRIAL_DAYS - daysSince).coerceAtLeast(0L).toInt()
            _trialDaysLeft.value = daysLeft
            updateIsPro()

            // Analytics
            if (isFirstRun) {
                AnalyticsProvider.instance.track(Event.TRIAL_STARTED,
                    mapOf("trial_days" to TRIAL_DAYS))
            }
            if (daysLeft == 0 && !_isRevenueCatPro.value) {
                AnalyticsProvider.instance.track(Event.TRIAL_EXPIRED)
            }
        }
        refresh()
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private fun CustomerInfo.hasPro(): Boolean =
        entitlements.active.containsKey(ENTITLEMENT_ID)

    private fun updateIsPro() {
        _isPro.value = _trialDaysLeft.value > 0 || _isRevenueCatPro.value
    }

    // ── Refresh desde RevenueCat ──────────────────────────────────────────────

    fun refresh() {
        Purchases.sharedInstance.getCustomerInfo(
            onSuccess = { customerInfo: CustomerInfo ->
                _isRevenueCatPro.value = customerInfo.hasPro()
                updateIsPro()
            },
            onError = { _: PurchasesError -> },
        )
        Purchases.sharedInstance.getOfferings(
            onSuccess = { offerings: Offerings -> _offerings.value = offerings },
            onError   = { _: PurchasesError -> },
        )
    }

    // ── Compra ────────────────────────────────────────────────────────────────

    fun purchase(
        rcPackage: Package,
        onSuccess: () -> Unit,
        onError:   (String) -> Unit,
    ) {
        _isLoading.value = true
        Purchases.sharedInstance.purchase(
            packageToPurchase = rcPackage,
            onError   = { _: PurchasesError, userCancelled: Boolean ->
                _isLoading.value = false
                if (!userCancelled) onError("Purchase failed")
            },
            onSuccess = { _: StoreTransaction, customerInfo: CustomerInfo ->
                _isRevenueCatPro.value = customerInfo.hasPro()
                _isLoading.value       = false
                updateIsPro()
                onSuccess()
            },
        )
    }

    // ── Debug / QA ────────────────────────────────────────────────────────────

    /**
     * SOLO DEBUG. Sobreescribe los días de trial restantes sin tocar DataStore.
     * Útil para verificar que el paywall, los lock icons y los banners se muestran
     * correctamente en cada estado del ciclo de vida del trial.
     *
     * Llamar desde los botones de debug de HomeScreen.
     */
    fun debugSetTrialDays(days: Int) {
        _trialDaysLeft.value = days
        updateIsPro()
    }

    // ── Restaurar ─────────────────────────────────────────────────────────────

    fun restore(
        onSuccess: () -> Unit,
        onError:   (String) -> Unit,
    ) {
        _isLoading.value = true
        Purchases.sharedInstance.restorePurchases(
            onSuccess = { customerInfo: CustomerInfo ->
                _isRevenueCatPro.value = customerInfo.hasPro()
                _isLoading.value       = false
                updateIsPro()
                AnalyticsProvider.instance.track(Event.PURCHASE_RESTORED)
                onSuccess()
            },
            onError = { _: PurchasesError ->
                _isLoading.value = false
                onError("Restore failed")
            },
        )
    }
}

File: apps/ios/WakeyWakey/Billing/EntitlementManager.swift
import Foundation
import RevenueCat
import UserNotifications

/// Gestiona el acceso Pro combinando dos fuentes:
///  1. Trial de 14 días desde la primera instalación (sin pedir nada)
///  2. Suscripción o compra única via RevenueCat / App Store
///
/// `isPro` = trial activo OR suscripción/compra activa
@MainActor
class EntitlementManager: ObservableObject {

    static let shared = EntitlementManager()

    static let entitlementID = "WakeyWakey Pro"
    static let trialDays     = 14

    private let installDateKey = "ww_install_date"

    // MARK: - Estado

    @Published var isPro:         Bool = true   // optimista hasta que RevenueCat responda
    @Published var trialDaysLeft: Int
    @Published var offerings:     Offerings?
    @Published var isLoading:     Bool = false
    @Published var restoreError:  String?

    private var isRevenueCatPro = false

    // MARK: - Init

    init() {
        let installDate = Self.getOrSetInstallDate()
        let elapsed     = Calendar.current.dateComponents([.day], from: installDate, to: Date()).day ?? 0
        trialDaysLeft   = max(0, Self.trialDays - elapsed)
        isPro           = trialDaysLeft > 0   // actualizado cuando RevenueCat responda
    }

    // MARK: - Configurar RevenueCat (llamar en App.init)

    func configure() {
        Purchases.logLevel = .error
        Purchases.configure(withAPIKey: "appl_taYwunKuycUTotfBfjloXeSwBAT")
        refresh()
    }

    // MARK: - Refresh desde RevenueCat

    func refresh() {
        Task {
            do {
                let info        = try await Purchases.shared.customerInfo()
                isRevenueCatPro = info.entitlements.active[Self.entitlementID] != nil
                updateIsPro()

                let off = try await Purchases.shared.offerings()
                offerings = off
            } catch {
                // Sin red: mantenemos el estado del trial
            }
        }
    }

    // MARK: - Compra

    func purchase(_ package: Package) async {
        isLoading = true
        defer { isLoading = false }
        do {
            let result      = try await Purchases.shared.purchase(package: package)
            isRevenueCatPro = result.customerInfo.entitlements.active[Self.entitlementID] != nil
            updateIsPro()
        } catch {
            // El usuario canceló o hubo error — no hacer nada
        }
    }

    // MARK: - Restaurar

    func restore() async {
        isLoading    = true
        restoreError = nil
        defer { isLoading = false }
        do {
            let info        = try await Purchases.shared.restorePurchases()
            isRevenueCatPro = info.entitlements.active[Self.entitlementID] != nil
            updateIsPro()
            if !isRevenueCatPro {
                restoreError = "No se encontró ninguna compra anterior asociada a este Apple ID."
            }
        } catch {
            restoreError = "No se pudo conectar con App Store. Inténtalo de nuevo."
        }
    }

    // MARK: - Helpers

    private func updateIsPro() {
        let wasProBefore = isPro
        #if DEBUG
        isPro = true
        return
        #endif
        isPro = trialDaysLeft > 0 || isRevenueCatPro

        if wasProBefore && !isPro {
            // Trial expirado o suscripción cancelada → cancelar notificaciones y limpiar widget
            UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
            WidgetDataWriter.clear()
        } else if !wasProBefore && isPro {
            // Acaba de suscribirse o restaurar → reprogramar notificaciones
            let settings = SettingsStore.shared
            let cal      = CalendarService.shared
            Task {
                await AlertScheduler.shared.rescheduleAll(
                    events:        cal.todayEvents,
                    minutesBefore: settings.alertMinutesBefore,
                    soundName:     settings.alertSoundName
                )
            }
        }
    }

    private static func getOrSetInstallDate() -> Date {
        let key = "ww_install_date"
        if let stored = UserDefaults.standard.object(forKey: key) as? Date {
            return stored
        }
        let now = Date()
        UserDefaults.standard.set(now, forKey: key)
        return now
    }

    // MARK: - Debug

    #if DEBUG
    func debugSetTrialDays(_ days: Int) {
        trialDaysLeft = days
        updateIsPro()
    }
    #endif
}
